packages/flutter_tools/lib/src/web/bootstrap.dart

```diff
@@ -216,11 +231,82 @@ let _scriptUrls = {
     window.\$dartLoader.loadConfig = loadConfig;
     window.\$dartLoader.loader = loader;
 
-    // TODO(srujzs): Support hot restart.
-
     // Begin loading libraries
     loader.nextAttempt();
-  }
+
+    // Set up stack trace mapper.
+    if (window.\$dartStackTraceUtility &&
+        !window.\$dartStackTraceUtility.ready) {
+      window.\$dartStackTraceUtility.ready = true;
+      window.\$dartStackTraceUtility.setSourceMapProvider(function(url) {
+        var baseUrl = window.location.protocol + '//' + window.location.host;
+        url = url.replace(baseUrl + '/', '');
+        if (url == 'dart_sdk.js') {
+          return dartDevEmbedder.debugger.getSourceMap('dart_sdk');
+        }
+        url = url.replace(".lib.js", "");
+        return dartDevEmbedder.debugger.getSourceMap(url);
+      });
+    }
+
+    let currentUri = _currentScript.src;
+    // We should have written a file containing all the scripts that need to be
+    // reloaded into the page. This is then read when a hot restart is triggered
+    // in DDC via the `\$dartReloadModifiedModules` callback.
+    let restartScripts = currentUri + '.restartScripts';
+
+    if (!window.\$dartReloadModifiedModules) {
+      window.\$dartReloadModifiedModules = (function(appName, callback) {
+        var xhttp = new XMLHttpRequest();
+        xhttp.withCredentials = true;
+        xhttp.onreadystatechange = function() {
+          // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
+          if (this.readyState == 4 && this.status == 200 || this.status == 304) {
+            var scripts = JSON.parse(this.responseText);
+            var numToLoad = 0;
+            var numLoaded = 0;
+            for (var i = 0; i < scripts.length; i++) {
+              var script = scripts[i];
+              if (script.id == null) continue;
+              var src = _currentDirectory + script.src.toString();
+              var oldSrc = window.\$dartLoader.moduleIdToUrl.get(script.id);
+              // Only compare the search parameters which contain the cache
+              // busting portion of the uri. The path might be different if the
+              // script is loaded from a different application on the page.
+              if (window.\$dartLoader.moduleIdToUrl.has(script.id) &&
+                  new URL(oldSrc).search == new URL(src).search) continue;
+
+              // We might actually load from a different uri, delete the old one
+              // just to be sure.
+              window.\$dartLoader.urlToModuleId.delete(oldSrc);
+
+              window.\$dartLoader.moduleIdToUrl.set(script.id, src);
+              window.\$dartLoader.urlToModuleId.set(src, script.id);
+
+              numToLoad++;
+
+              var el = document.getElementById(script.id);
+              if (el) el.remove();
```

Comment:
Seems like this is removing a script from the page but do we know for sure that it is no longer needed? I wonder what happens after a hot reload if you step into old code from a previous generation that is still kept alive through some existing state in the app but the script is gone.

Reply:
Hot restart reinitializes the state so I don't think such a scenario is possible unless you go through interop, but we don't really plan on handling that scenario.